storage/rocksdb/ha_rocksdb.cc

```diff
@@ -6134,10 +6225,12 @@ class Rdb_writebatch_impl : public Rdb_transaction {
     }
     assert_snapshot_invariants();
     if (has_snapshot(table_type)) {
-      assert(statement_snapshot_type == snapshot_type::CURRENT);
-      statement_snapshot_type = snapshot_type::NONE;
+      assert(get_statement_snapshot_type() == snapshot_type::CURRENT);
+      statement_snapshot_type.store(snapshot_type::NONE,
+                                    std::memory_order_relaxed);
       rdb->ReleaseSnapshot(m_read_opts[table_type].snapshot);
       m_read_opts[table_type].snapshot = nullptr;
```

Comment:
For line 6231 and line 6232, may also not safe for other threads(use after delete issue)

current thread finished execute "rdb->ReleaseSnapshot()" but before execute "m_read_opts[table_type].snapshot = nullptr", while another thread is calling tx->get_snapshot_ts();(https://github.com/facebook/mysql-5.6/blob/fb-mysql-8.0.32/storage/rocksdb/ha_rocksdb.cc#L3987). the snapshot has been cleared but m_read_opts[USER_TABLE].snapshot may still reference old memory


maybe change to

```
      m_read_opts[table_type].snapshot = nullptr; 
      rdb->ReleaseSnapshot(m_read_opts[table_type].snapshot);
```

Reply:
`get_snapshot_ts` does not race here, because this PR introduces a caching field exactly for this scenario.

Having said that, `m_read_opts[table_type].snapshot` is indeed accessed incorrectly in other contexts (see my e-mail) and will be fixed one way or another in a follow-up